src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs

```diff
@@ -1012,128 +978,76 @@ private bool IsMatchingAssembly(PEAssemblySymbol peAssembly)
                         {
-                            refs.Add(sourceReferencedAssemblies[i]);
+                            result.Add(sourceReferencedAssemblies[i]);
                         }
                     }
 
-                    for (i = 1; i < mCount; i++)
+                    for (int i = 1; i < modules.Length; i++)
                     {
-                        refs.AddRange(modules[i].GetReferencedAssemblies());
+                        result.AddRange(modules[i].GetReferencedAssemblies());
                     }
 
-                    referencedAssemblies = refs.ToImmutableAndFree();
+                    return result.ToImmutableAndFree();
                 }
 
                 internal override AssemblySymbol CreateAssemblySymbol()
                 {
-                    return new Symbols.Retargeting.RetargetingAssemblySymbol(_compilation.SourceAssembly, this.IsLinked);
+                    return new RetargetingAssemblySymbol(Compilation.SourceAssembly, this.IsLinked);
                 }
 
                 protected override void AddAvailableSymbols(List<AssemblySymbol> assemblies)
                 {
-                    assemblies.Add(_compilation.Assembly);
+                    assemblies.Add(Compilation.Assembly);
 
                     // accessing cached symbols requires a lock
                     lock (SymbolCacheAndReferenceManagerStateGuard)
                     {
-                        _compilation.AddRetargetingAssemblySymbolsNoLock(assemblies);
+                        Compilation.AddRetargetingAssemblySymbolsNoLock(assemblies);
                     }
                 }
 
                 public override bool IsMatchingAssembly(AssemblySymbol candidateAssembly)
                 {
-                    var retargeting = candidateAssembly as Symbols.Retargeting.RetargetingAssemblySymbol;
-                    AssemblySymbol asm;
-
-                    if ((object)retargeting != null)
-                    {
-                        asm = retargeting.UnderlyingAssembly;
-                    }
-                    else
-                    {
-                        asm = candidateAssembly as SourceAssemblySymbol;
-                    }
-
-                    Debug.Assert(!(asm is Symbols.Retargeting.RetargetingAssemblySymbol));
-
-                    return ReferenceEquals(asm, _compilation.Assembly);
-                }
-
-                public override bool ContainsNoPiaLocalTypes
-                {
-                    get
-                    {
-                        return _compilation.MightContainNoPiaLocalTypes();
-                    }
-                }
-
-                public override bool DeclaresTheObjectClass
-                {
-                    get
-                    {
-                        return _compilation.DeclaresTheObjectClass;
-                    }
+                    var asm = (candidateAssembly as RetargetingAssemblySymbol)?.UnderlyingAssembly ?? candidateAssembly;
```

Comment:
The logic is not equivalent to the original. Please revert code back to the 'if' statement.